File: Airship/AirshipCore/Source/common/UARetailEventTemplate.m
/* Copyright Airship and Contributors */

#import "UARetailEventTemplate.h"
#import "UAirship.h"
#import "UAAnalytics.h"
#import "UACustomEvent+Internal.h"

#define kUARetailEventTemplate @"retail"
#define kUABrowsedProductEvent @"browsed"
#define kUAAddedToCartEvent @"added_to_cart"
#define kUAStarredProductEvent @"starred_product"
#define kUASharedProductEvent @"shared_product"
#define kUAPurchasedEvent @"purchased"
#define kUARetailEventTemplateLifetimeValue @"ltv"
#define kUARetailEventTemplateIdentifier @"id"
#define kUARetailEventTemplateCategory @"category"
#define kUARetailEventTemplateDescription @"description"
#define kUARetailEventTemplateBrand @"brand"
#define kUARetailEventTemplateNewItem @"new_item"
#define kUARetailEventTemplateSource @"source"
#define kUARetailEventTemplateMedium @"medium"

@interface UARetailEventTemplate()
@property (nonatomic, copy) NSString *eventName;
@property (nonatomic, copy) NSString *source;
@property (nonatomic, copy) NSString *medium;
@property (nonatomic, assign) BOOL newItemSet;
@end

@implementation UARetailEventTemplate

- (instancetype)initWithName:(NSString *)name
                   withValue:(NSDecimalNumber *)eventValue
                  withSource:(NSString *)source
                  withMedium:(NSString *)medium {
    self = [super init];
    if (self) {
        self.eventName = name;
        self.eventValue = eventValue;
        self.source = source;
        self.medium = medium;
    }

    return self;
}

+ (instancetype)browsedTemplate {
    return [self browsedTemplateWithValue:nil];
}

+ (instancetype)browsedTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self browsedTemplateWithValue:decimalValue];
}

+ (instancetype)browsedTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUABrowsedProductEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)addedToCartTemplate {
    return [self addedToCartTemplateWithValue:nil];
}

+ (instancetype)addedToCartTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self addedToCartTemplateWithValue:decimalValue];
}

+ (instancetype)addedToCartTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAAddedToCartEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)starredProductTemplate {
    return [self starredProductTemplateWithValue:nil];
}

+ (instancetype)starredProductTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self starredProductTemplateWithValue:decimalValue];
}

+ (instancetype)starredProductTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAStarredProductEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)purchasedTemplate {
    return [self purchasedTemplateWithValue:nil];
}

+ (instancetype)purchasedTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self purchasedTemplateWithValue:decimalValue];
}

+ (instancetype)purchasedTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAPurchasedEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)sharedProductTemplate {
    return [self sharedProductTemplateWithValue:nil];
}

+ (instancetype)sharedProductTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self sharedProductTemplateWithValue:decimalValue];
}

+ (instancetype)sharedProductTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)sharedProductTemplateWithSource:(NSString *)source
                                  withMedium:(NSString *)medium {
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:nil
                           withSource:source
                           withMedium:medium];
}

+ (instancetype)sharedProductTemplateWithValueFromString:(NSString *)eventValue
                                           withSource:(NSString *)source
                                           withMedium:(NSString *)medium {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:decimalValue
                           withSource:source
                           withMedium:medium];
}

+ (instancetype)sharedProductTemplateWithValue:(NSDecimalNumber *)eventValue
                                 withSource:(NSString *)source
                                 withMedium:(NSString *)medium {
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:eventValue
                           withSource:source
                           withMedium:medium];
}

- (void)setEventValue:(NSDecimalNumber *)eventValue {
    if (!eventValue) {
        _eventValue = nil;
    } else {
        if ([eventValue isKindOfClass:[NSDecimalNumber class]]) {
            _eventValue = eventValue;
        } else {
            _eventValue = [NSDecimalNumber decimalNumberWithDecimal:[eventValue decimalValue]];
        }
    }
}

- (void)setIsNewItem:(BOOL)isNewItem {
    self.newItemSet = YES;
    _isNewItem = isNewItem;
}

- (UACustomEvent *)createEvent {
    UACustomEvent *event = [UACustomEvent eventWithName:self.eventName];

    NSMutableDictionary *propertyDictionary = [NSMutableDictionary dictionary];
    
    if (self.eventValue) {
        [event setEventValue:self.eventValue];
    }

    if (self.eventValue && [self.eventName isEqualToString:kUAPurchasedEvent]) {
        [propertyDictionary setValue:@YES forKey:kUARetailEventTemplateLifetimeValue];
    } else {
       [propertyDictionary setValue:@NO forKey:kUARetailEventTemplateLifetimeValue];
    }

    if (self.transactionID) {
        [event setTransactionID:self.transactionID];
    }

    if (self.identifier) {
        [propertyDictionary setValue:self.identifier forKey:kUARetailEventTemplateIdentifier];
    }

    if (self.category) {
        [propertyDictionary setValue:self.category forKey:kUARetailEventTemplateCategory];
    }

    if (self.eventDescription) {
        [propertyDictionary setValue:self.eventDescription forKey:kUARetailEventTemplateDescription];
    }

    if (self.brand) {
        [propertyDictionary setValue:self.brand forKey:kUARetailEventTemplateBrand];
    }

    if (self.newItemSet) {
        [propertyDictionary setValue:@(self.isNewItem) forKey:kUARetailEventTemplateNewItem];
    }
    if (self.source) {
        [propertyDictionary setValue:self.source forKey:kUARetailEventTemplateSource];
    }

    if (self.medium) {
        [propertyDictionary setValue:self.medium forKey:kUARetailEventTemplateMedium];
    }

    event.templateType = kUARetailEventTemplate;
    event.properties = propertyDictionary;
    
    return event;
}

@end

File: Airship/AirshipWrapper/Analytics/Events/Templates/UARetailEventTemplate.swift
/* Copyright Airship and Contributors */

import Foundation
public import AirshipCore

/// This singleton provides an interface to the functionality provided by the Airship iOS Push API.
@objc
public class UARetailEventTemplate: NSObject {
    
    private var template: RetailEventTemplate
    
    /**
     * The event's value. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     */
    @objc
    public var eventValue: NSNumber? {
        get {
           return template.eventValue
        }
        set {
            template.eventValue = newValue
        }
    }

    /**
     * The event's transaction ID. The transaction ID's length must not exceed 255
     * characters or it will invalidate the event.
     */
    @objc
    public var transactionID: String? {
        get {
           return template.transactionID
        }
        set {
            template.transactionID = newValue
        }
    }

    /**
     * The event's ID.
     */
    @objc
    public var identifier: String? {
        get {
           return template.identifier
        }
        set {
            template.identifier = newValue
        }
    }

    /**
     * The event's category.
     */
    @objc
    public var category: String? {
        get {
           return template.category
        }
        set {
            template.category = newValue
        }
    }

    /**
     * The event's description.
     */
    @objc
    public var eventDescription: String? {
        get {
            return template.eventDescription
        }
        set {
            template.eventDescription = newValue
        }
    }

    /**
     * The brand..
     */
    @objc
    public var brand: String? {
        get {
           return template.brand
        }
        set {
            template.brand = newValue
        }
    }

    /**
     * If the item is new or not.
     */
    @objc
    public var isNewItem: Bool {
        get {
            return template.isNewItem
        }
        set {
            template.isNewItem = newValue
        }
    }
    
    @objc
    public init(template: RetailEventTemplate) {
        self.template = template
    }

    /**
     * Factory method for creating a browsed event template.
     * - Returns: A Retail event template instance
     */
    @objc(browsedTemplate)
    public class func browsedTemplate() -> UARetailEventTemplate {
        return browsedTemplate(value: nil)
    }

    /**
     * Factory method for creating a browsed event template with a value.
     *
     * - Parameter valueString: The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(browsedTemplateWithValueFromString:)
    public class func browsedTemplate(valueString: String?)
        -> UARetailEventTemplate
    {
        let decimalValue =
            valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return browsedTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating a browsed event template with a value.
     *
     * - Parameter value: The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(browsedTemplateWithValue:)
    public class func browsedTemplate(value: NSNumber?) -> UARetailEventTemplate {
        let template = RetailEventTemplate("browsed", value: value)
        return UARetailEventTemplate(template: template)
    }

    /**
     * Factory method for creating an addedToCart event template.
     * - Returns: A Retail event template instance
     */
    @objc(addedToCartTemplate)
    public class func addedToCartTemplate() -> UARetailEventTemplate {
        return addedToCartTemplate(value: nil)
    }

    /**
     * Factory method for creating an addedToCart event template with a value.
     *
     * - Parameter valueString: The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(addedToCartTemplateWithValueFromString:)
    public class func addedToCartTemplate(valueString: String?)
        -> UARetailEventTemplate
    {
        let decimalValue =
            valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return addedToCartTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating an addedToCart event template with a value.
     *
     * - Parameter value: The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(addedToCartTemplateWithValue:)
    public class func addedToCartTemplate(value: NSNumber?)
        -> UARetailEventTemplate
    {
        let template = RetailEventTemplate("added_to_cart", value: value)
        return UARetailEventTemplate(template: template)
    }

    /**
     * Factory method for creating a starredProduct event template
     * - Returns: A Retail event template instance
     */
    @objc(starredProductTemplate)
    public class func starredProductTemplate() -> UARetailEventTemplate {
        return starredProductTemplate(value: nil)
    }

    /**
     * Factory method for creating a starredProduct event template with a value.
     *
     * - Parameter valueString: The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(starredProductTemplateWithValueFromString:)
    public class func starredProductTemplate(valueString: String?)
        -> UARetailEventTemplate
    {
        let decimalValue =
            valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return starredProductTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating a starredProduct event template with a value.
     *
     * - Parameter value: The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(starredProductTemplateWithValue:)
    public class func starredProductTemplate(value: NSNumber?)
        -> UARetailEventTemplate
    {
        let template = RetailEventTemplate("starred_product", value: value)
        return UARetailEventTemplate(template: template)
    }

    /**
     * Factory method for creating a purchased event template.
     * - Returns: A Retail event template instance
     */
    @objc(purchasedTemplate)
    public class func purchasedTemplate() -> UARetailEventTemplate {
        return purchasedTemplate(value: nil)
    }

    /**
     * Factory method for creating a purchased event template with a value.
     *
     * - Parameter valueString: The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(purchasedTemplateWithValueFromString:)
    public class func purchasedTemplate(valueString: String?)
        -> UARetailEventTemplate
    {
        let decimalValue =
            valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return purchasedTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating a purchased event template with a value.
     *
     * - Parameter value: The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(purchasedTemplateWithValue:)
    public class func purchasedTemplate(value: NSNumber?) -> UARetailEventTemplate
    {
        let template = RetailEventTemplate("purchased", value: value)
        return UARetailEventTemplate(template: template)
    }

    /**
     * Factory method for creating a sharedProduct template event.
     * - Returns: A Retail event template instance
     */
    @objc(sharedProductTemplate)
    public class func sharedProductTemplate() -> UARetailEventTemplate {
        return sharedProductTemplate(value: nil, source: nil, medium: nil)
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * - Parameter valueString: The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(sharedProductTemplateWithValueFromString:)
    public class func sharedProductTemplate(valueString: String?)
        -> UARetailEventTemplate
    {
        let decimalValue =
            valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return sharedProductTemplate(
            value: decimalValue,
            source: nil,
            medium: nil
        )
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * - Parameter value: The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * - Returns: A Retail event template instance
     */
    @objc(sharedProductTemplateWithValue:)
    public class func sharedProductTemplate(value: NSNumber?)
        -> UARetailEventTemplate
    {
        return sharedProductTemplate(value: value, source: nil, medium: nil)
    }

    /**
     * Factory method for creating a sharedProduct event template.
     * - Parameter source: The source as an NSString.
     * - Parameter medium: The medium as an NSString
     * - Returns: A Retail event template instance.
     */
    @objc(sharedProductTemplateWithSource:withMedium:)
    public class func sharedProductTemplate(source: String?, medium: String?)
        -> UARetailEventTemplate
    {
        return sharedProductTemplate(value: nil, source: source, medium: medium)
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * - Parameter valueString: The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * - Parameter source: The source as an NSString.
     * - Parameter medium: The medium as an NSString.
     * - Returns: A Retail event template instance
     */
    @objc(sharedProductTemplateWithValueFromString:withSource:withMedium:)
    public class func sharedProductTemplate(
        valueString: String?,
        source: String?,
        medium: String?
    ) -> UARetailEventTemplate {
        let decimalValue =
            valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return sharedProductTemplate(
            value: decimalValue,
            source: source,
            medium: medium
        )
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * - Parameter value: The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * - Parameter source: The source as an NSString.
     * - Parameter medium: The medium as an NSString.
     * - Returns: A Retail event template instance
     */
    @objc(sharedProductTemplateWithValue:withSource:withMedium:)
    public class func sharedProductTemplate(
        value: NSNumber?,
        source: String?,
        medium: String?
    ) -> UARetailEventTemplate {
        let template = RetailEventTemplate(
            "shared_product",
            value: value,
            source: source,
            medium: medium
        )
        return UARetailEventTemplate(template: template)
    }

    /**
     * Factory method for creating a wishlist event template.
     * - Returns: A Retail event template instance
     */
    @objc(wishlistTemplate)
    public class func wishlistTemplate() -> UARetailEventTemplate {
        return wishlistTemplate(name: nil, wishlistID: nil)
    }

    /**
     * Factory method for creating a wishlist event template with a wishlist name and ID.
     *
     * - Parameter name: The name of the wishlist as as string.
     * - Parameter wishlistID: The ID of the wishlist as as string.
     * - Returns: A Retail event template instance
     */
    @objc(wishlistTemplateWithName:wishlistID:)
    public class func wishlistTemplate(name: String?, wishlistID: String?)
        -> UARetailEventTemplate
    {
        let template = RetailEventTemplate(
            "wishlist",
            wishlistName: name,
            wishlistID: wishlistID
        )
        return UARetailEventTemplate(template: template)
    }

    /**
     * Creates the custom media event.
     */
    @objc
    public func createEvent() -> CustomEvent {
        self.template.createEvent()
    }
    
}
